docs/proposals/20220927-label-sync-between-machine-and-nodes.md

```diff
@@ -0,0 +1,137 @@
+---
+
+title: Label Sync Between Machine and underlying Kubernetes Nodes
+authors:
+- "@arvinderpal" (original proposal author)
+- "@enxebre"     (original proposal author)
+- "@fabriziopandini"
+reviewers:
+- @sbueringer
+- @oscar
+- @vincepri
+creation-date: 2022-02-10
+last-updated: 2022-02-26
+status: implementable
+replaces:
+superseded-by:
+
+---
+
+# Label Sync Between Machine and underlying Kubernetes Nodes
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+This document discusses how labels placed on a Machine can kept in sync with the corresponding Kubernetes Node.
+
+## Motivation
+
+Managing labels on Kubernetes nodes has been a long standing [issue](https://github.com/kubernetes-sigs/cluster-api/issues/493) in Cluster API.
+
+The following challenges have been identified through various iterations:
+
+- Define how labels propagate from Machine to node.
+- Define how labels propagate from ClusterClass to KubeadmControlPlane/MachineDeployments/Machine Pools, and ultimately to Machines.
+- Define how to prevent that label propagation triggers unnecessary rollouts.
+
+With the "divide and conquer" principle in mind this proposal aims to address the first point only, while the remaining points are going to be addressed in separated, complementary efforts.
+
+### Goals
+
+- Support label sync from Machine to the linked Kubernetes node, limited to `node-role.kubernetes.io/` prefix and the `node-restriction.kubernetes.io` domain.
+- Support syncing labels from Machine to the linked Kubernetes node for the Cluster API owned `node.cluster.x-k8s.io` domain.
+
+### Non-Goals
+
+- Support for arbitrary/user-specified label prefixes.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin/user, I would like a declarative and secure means by which to assign roles to my nodes. For example, when I do `kubectl get nodes`, I want the ROLES column to list my assigned roles for the nodes.
+
+#### Story 2
+
+As a cluster admin/user, for the purpose of workload placement, I would like a declarative and secure means by which I can add/remove labels on groups of nodes. For example, I want to run my outward facing nginx pods on a specific set of nodes (e.g. md-secure-zone) to reduce exposure to the company's publicly trusted server certificates.
+
+#### Story 3
+
+As a cluster admin/user, I want that Cluster API label management on Kubernetes Nodes doesn't conflict with labels directly managed by users or by other controllers.
+
+### Implementation Details/Notes/Constraints
+
+While designing a solution for syncing labels between Machine and underlying Kubernetes Nodes two main concerns have been considered:
+
+- Security, because Node labels can be used to schedule and/or limit workloads to a predetermined set of nodes.
+- Impact on other components applying labels on Kubernetes Nodes, like e.g. Kubeadm, CPI etc.
+
+### Label Prefixes
```

Comment:
Nit: Should this be just Labels or Labels domains & prefixes now? (given we only have one prefix now)

Reply:
done